src/target/riscv/riscv.c

```diff
@@ -2266,120 +2354,95 @@ int sample_memory(struct target *target)
 int riscv_openocd_poll(struct target *target)
 {
 	LOG_DEBUG("polling all harts");
-	enum target_state old_state = target->state;
+
+	struct list_head *targets;
+
+	LIST_HEAD(single_target_list);
+	struct target_list single_target_entry = {
+		.lh = {NULL, NULL},
+		.target = target
+	};
 
 	if (target->smp) {
-		unsigned should_remain_halted = 0;
-		unsigned should_resume = 0;
-		struct target_list *list;
-		foreach_smp_target(list, target->smp_targets) {
-			struct target *t = list->target;
-			if (!target_was_examined(t))
-				continue;
-			enum riscv_poll_hart out = riscv_poll_hart(t, t->coreid);
-			switch (out) {
-			case RPH_NO_CHANGE:
-				break;
-			case RPH_DISCOVERED_RUNNING:
-				t->state = TARGET_RUNNING;
-				t->debug_reason = DBG_REASON_NOTHALTED;
-				break;
-			case RPH_DISCOVERED_HALTED:
-				t->state = TARGET_HALTED;
-				enum riscv_halt_reason halt_reason =
-					riscv_halt_reason(t);
-				if (set_debug_reason(t, halt_reason) != ERROR_OK)
-					return ERROR_FAIL;
+		targets = target->smp_targets;
+	} else {
+		/* Make a list that just contains a single target, so we can
+		 * share code below. */
+		list_add(&single_target_entry.lh, &single_target_list);
+		targets = &single_target_list;
+	}
 
-				if (halt_reason == RISCV_HALT_BREAKPOINT) {
-					int retval;
-					switch (riscv_semihosting(t, &retval)) {
-					case SEMI_NONE:
-					case SEMI_WAITING:
-						/* This hart should remain halted. */
-						should_remain_halted++;
-						break;
-					case SEMI_HANDLED:
-						/* This hart should be resumed, along with any other
-							 * harts that halted due to haltgroups. */
-						should_resume++;
-						break;
-					case SEMI_ERROR:
-						return retval;
-					}
-				} else if (halt_reason != RISCV_HALT_GROUP) {
-					should_remain_halted++;
-				}
-				break;
+	unsigned should_remain_halted = 0;
+	unsigned should_resume = 0;
+	unsigned halted = 0;
+	unsigned running = 0;
+	struct target_list *entry;
+	foreach_smp_target(entry, targets) {
+		struct target *t = entry->target;
+		riscv_info_t *info = riscv_info(t);
 
-			case RPH_ERROR:
-				return ERROR_FAIL;
-			}
-		}
+		info->needs_callback = false;
```

Comment:
This assert would work with the change suggested at line 2434.

```suggestion
		/* Any pending callback for the target must have been processed already. */
		assert(!info->needs_callback);
```

Reply:
I don't like the assert. It's simple to clear the state before it might get set. It's complicated to track every possible error case and make sure the state is cleared if it was set and we didn't get to actually send the callback. On top of that we're unlikely to catch every possible of those error cases in a test, so it seems likely the code is incomplete and a user will encounter an assert that kills their debug session for a fairly innocuous reason.

Reply:
`info->needs_callback == false` is an invariant at this point in the code, and that is exactly what `assert()`s are for :)

In any case, please feel free to ignore this suggestion. It is not really important to have the assert here.